### DISCUSSION TITLE
Designing different levels of Nginx and Traefik Proxies?  App-based vs. Ingress

### DISCUSSION THREAD

**Discussion by al-yakubovich** [BretFisher/ama#118] [Q&A]
Hello Bret! I got confused about what technologies I need to use on top of Flask/FastAPI if I want to serve a ML model. I hope you can help me. 

I am looking to the following stacks: 
`I. Flask + uWSGI`
`II. FastAPI + Uvicorn with Gunicorn`

On top of that I can put Nginx as a reverse proxy (load balancing, caching, security, etc.):
`I. Flask + uWSGI + Nginx`
`II. FastAPI + Uvicorn/Gunicorn + Nginx`

My questions:

1. Do I need to use Traefik on top of that? Or do I need to change Nginx to Traefik if I want to use Traefik?

2. If I have an application which only accepts POST requests is is recommended to use Nginx and/or Traefik on top of FastAPI + Uvicorn/Gunicorn?

3. If I will use Tensorflow Serving or other ML serving solutions (Kubeflow, MLflow, Seldon, etc) is it still recommended to wrap up Tensorflow Serving into FastAPI + Uvicorn/Gunicorn + Nginx or/and Traefik?

Thank you!

**Comment by BretFisher** [comment]
Hey there. I'm not a ML person, so my answers will be more generic:

* Your "ingress proxy" sits in front of *all* your apps in a cluser and routes traffic to the correct container based on HTTP rules like DNS name or URL path, etc. This proxy is HTTP specific, since it routes based on HTTP headers.
* Nginx can be a web server, a file-caching proxy, and a HTTP-to-Socket gateway (packets come in as HTTP and Nginx turns those into app-specific packets like fastcgi, WSGI, etc.
* Traefik doesn't do any of that. It's a HTTP (and now other TCP protocols) proxy for routing. It doesn't cache, translate to non HTTP/S endpoints, or act as a web server for your apps.
* I think of these different layers as separate parts of the stack. Your app should be built (in one or more containers) to run independently on its own server. So maybe in your case, because you want Gunicorn to run the app, and Nginx maybe to cache static assets (I have no opinion there on what's best or more performant). *Then*, when you want to run multiple apps on the same server (or set of servers in a cluster), you would *add* a Ingress Routing Proxy (sometimes called a Gateway) to act as the gatekeeper on incoming HTTP/S connections. It has rules that route traffic to different apps/containers/pods and that's it.
* Most Swarm/Kubernetes clusters will have a Ingress Contoroller like Traefik/Nginx/Envoy in them. That's a infrastructure component that's for the whole cluster. Your individual apps might be designed for a single K8s Pod with one Gunicorn Pod and one Nginx Pod to work together (I see this often with PHP-FPM + Nginx)... but that Nginx would be specific for that one Pod/app, so it's serving a different purpose then the whole-cluster-ingress-proxy.  So don't think of them as redundant.
* Yes, in the end, a Swarm/K8s cluster may have many proxies. We joke that K8s is "just a bunch of proxies and a few apps". Welcome to a cloud native microservice world 😸 

I hope that helps, and good luck!
